sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -589,6 +594,15 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
       IntervalUtils.castStringToDTInterval(s, it.startField, it.endField))
     case _: DayTimeIntervalType => buildCast[Long](_, s =>
       IntervalUtils.durationToMicros(IntervalUtils.microsToDuration(s), it.endField))
+    case x: LongType =>
+      assert(it.startField == it.endField)
+      b => IntervalUtils.longToDayTimeInterval(
+        x.exactNumeric.asInstanceOf[Numeric[Any]].toLong(b), it.endField)
+    case x: IntegralType =>
+      assert(it.startField == it.endField)
+      assert(x != LongType)
+      b => IntervalUtils.intToDayTimeInterval(
+        x.integral.asInstanceOf[Integral[Any]].toInt(b), it.endField)
```

Comment:
Merge byte/short/int into one case branch to reduce the code and add `assert(x != LongType)` although I think it won't get here for LongType

Reply:
seems better to unify the code more
```
case x: IntegralType =>
  assert(it.startField == it.endField)
  if (x == LongType) ... else ...
```

Reply:
done